`mvn install` fails on multi-module maven project with more than one itest

I have a maven project with two modules containing one integration test each. itests are based on bnd-testing-maven-plugin.

* If I `mvn install` from the modules folders, each itest completes successfully.
* If I `mvn install`from the parent pom folder, the second itest fails

I tried to setup [a maven project](https://github.com/mrulli/com.flairbit.examples.bndmultipleitest) to reproduce the issue. 

Sorry for the tests: they are quite nonsensical but I did not have time to think about something better. Anyway this is not important to reproduce the issue: also an empty test would do.

I have added tests to the `bnd-testing-maven-plugin` to confirm that this isn't an issue on our end. Having tried the tests in your example repository, the first exception I saw was a `NullPointerException` from deep inside Equinox. I switched to using Felix, and the test progressed much further, but still encountered an error in setup.

The exception gets pretty garbled, but the following is visible:

    Caused by: java.lang.ClassNotFoundException: com.sun.xml.bind.DatatypeConverterImpl not found by org.apache.karaf.features.core [24]
	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1574)
	at org.apache.felix.framework.BundleWiringImpl.access$400(BundleWiringImpl.java:79)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2018)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 43 more
     found by org.apache.karaf.features.core [24]
	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1574)
	at org.apache.felix.framework.BundleWiringImpl.access$400(BundleWiringImpl.java:79)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2018)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 43 more

At this point I'm pretty confident that the plugin is launching things correctly, and that there's something unpleasant going on in the setup of the test.

--- a/maven/bnd-testing-maven-plugin/.settings/org.eclipse.core.resources.prefs
+++ b/maven/bnd-testing-maven-plugin/.settings/org.eclipse.core.resources.prefs
@@ -2,4 +2,5 @@ eclipse.preferences.version=1
 encoding//src/main/java=UTF-8
 encoding//src/main/resources=UTF-8
 encoding//src/test/java=UTF-8
+encoding//src/test/resources=UTF-8
 encoding/<project>=UTF-8

--- a/maven/bnd-testing-maven-plugin/pom.xml
+++ b/maven/bnd-testing-maven-plugin/pom.xml
@@ -79,7 +79,7 @@
                     <localRepositoryPath>${project.build.directory}/integration-test/repo</localRepositoryPath>
                     <postBuildHookScript>verify.groovy</postBuildHookScript>
                     <goals>
-                        <goal>package</goal>
+                        <goal>integration-test</goal>
                     </goals>
                     <debug>true</debug>
                     <scriptVariables>

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/settings.xml
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/settings.xml
@@ -1,0 +1,15 @@
+<settings xmlns="http://maven.apache.org/SETTINGS/1.0.0"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0
+                          http://maven.apache.org/xsd/settings-1.0.0.xsd">
+	<localRepository />
+	<interactiveMode>false</interactiveMode>
+	<usePluginRegistry />
+	<offline />
+	<pluginGroups />
+	<servers />
+	<mirrors />
+	<proxies />
+	<profiles/>	
+	<activeProfiles />
+</settings>

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/pom.xml
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/pom.xml
@@ -1,0 +1,122 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+	<groupId>biz.aQute.bnd-test</groupId>
+	<artifactId>test</artifactId>
+	<version>0.0.1</version>
+	<packaging>pom</packaging>
+
+	<distributionManagement>
+		<repository>
+			<id>Test Repo</id>
+			<url>file:@project.build.directory@/integration-test/test-repo</url>
+		</repository>
+	</distributionManagement>
+
+	<modules>
+		<module>test-no-resolve</module>
+		<module>test-with-resolve</module>
+	</modules>
+
+	<dependencyManagement>
+		<dependencies>
+			<dependency>
+				<groupId>org.osgi</groupId>
+				<artifactId>org.osgi.core</artifactId>
+				<version>6.0.0</version>
+			</dependency>
+			<dependency>
+				<groupId>org.osgi</groupId>
+				<artifactId>osgi.enroute.junit.wrapper</artifactId>
+				<version>4.12.0</version>
+			</dependency>
+			<dependency>
+				<groupId>org.osgi</groupId>
+				<artifactId>osgi.enroute.hamcrest.wrapper</artifactId>
+				<version>1.3.0</version>
+			</dependency>
+			<dependency>
+				<groupId>org.apache.felix</groupId>
+				<artifactId>org.apache.felix.framework</artifactId>
+				<version>5.6.1</version>
+				<scope>runtime</scope>
+			</dependency>
+			<dependency>
+				<groupId>org.apache.felix</groupId>
+				<artifactId>org.apache.felix.scr</artifactId>
+				<version>2.0.2</version>
+				<scope>runtime</scope>
+			</dependency>
+		</dependencies>
+	</dependencyManagement>
+
+	<build>
+		<pluginManagement>
+			<plugins>
+				<plugin>
+					<groupId>biz.aQute.bnd</groupId>
+					<artifactId>bnd-testing-maven-plugin</artifactId>
+					<version>@project.version@</version>
+					<configuration>
+						<bndruns>
+							<bndrun>test.bndrun</bndrun>
+						</bndruns>
+					</configuration>
+					<executions>
+						<execution>
+							<id>testing</id>
+							<goals>
+								<goal>testing</goal>
+							</goals>
+						</execution>
+					</executions>
+				</plugin>
+				<plugin>
+					<groupId>biz.aQute.bnd</groupId>
+					<artifactId>bnd-indexer-maven-plugin</artifactId>
+					<version>@project.version@</version>
+					<configuration>
+						<includeJar>true</includeJar>
+						<localURLs>REQUIRED</localURLs>
+					</configuration>
+					<executions>
+						<execution>
+							<id>index</id>
+							<goals>
+								<goal>index</goal>
+							</goals>
+						</execution>
+					</executions>
+				</plugin>
+				<plugin>
+					<groupId>biz.aQute.bnd</groupId>
+					<artifactId>bnd-maven-plugin</artifactId>
+					<version>@project.version@</version>
+					<executions>
+						<execution>
+							<id>default-bnd-process</id>
+							<goals>
+								<goal>bnd-process</goal>
+							</goals>
+						</execution>
+					</executions>
+				</plugin>
+
+				<!-- The following configuration is required because bnd-maven-plugin 
+					generates the manifest to target/classes/META-INF/MANIFEST.MF but the normal 
+					maven-jar-plugin does not use it. If the jar-plugin is patched to pick up 
+					the manifest from this location, then the following config is not needed. -->
+				<plugin>
+					<groupId>org.apache.maven.plugins</groupId>
+					<artifactId>maven-jar-plugin</artifactId>
+					<version>2.6</version>
+					<configuration>
+						<archive>
+							<manifestFile>${project.build.outputDirectory}/META-INF/MANIFEST.MF</manifestFile>
+						</archive>
+					</configuration>
+				</plugin>
+			</plugins>
+		</pluginManagement>
+	</build>
+</project>

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-no-resolve/bnd.bnd
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-no-resolve/bnd.bnd
@@ -1,0 +1,1 @@
+Test-Cases: ${classes;CONCRETE;PUBLIC;NAMED;*Test}

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-no-resolve/pom.xml
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-no-resolve/pom.xml
@@ -1,0 +1,50 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+
+	<parent>
+		<groupId>biz.aQute.bnd-test</groupId>
+		<artifactId>test</artifactId>
+		<version>0.0.1</version>
+	</parent>
+
+	<groupId>biz.aQute.bnd-test</groupId>
+	<artifactId>test-no-resolve</artifactId>
+	<packaging>jar</packaging>
+
+	<dependencies>
+		<dependency>
+			<groupId>org.osgi</groupId>
+			<artifactId>org.osgi.core</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.osgi</groupId>
+			<artifactId>osgi.enroute.junit.wrapper</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.osgi</groupId>
+			<artifactId>osgi.enroute.hamcrest.wrapper</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.apache.felix</groupId>
+			<artifactId>org.apache.felix.framework</artifactId>
+		</dependency>
+	</dependencies>
+
+	<build>
+		<plugins>
+			<plugin>
+				<groupId>biz.aQute.bnd</groupId>
+				<artifactId>bnd-testing-maven-plugin</artifactId>
+			</plugin>
+			<plugin>
+				<groupId>biz.aQute.bnd</groupId>
+				<artifactId>bnd-indexer-maven-plugin</artifactId>
+			</plugin>
+			<plugin>
+				<groupId>biz.aQute.bnd</groupId>
+				<artifactId>bnd-maven-plugin</artifactId>
+			</plugin>
+		</plugins>
+	</build>
+</project>

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-no-resolve/src/main/java/bnd/test/SimpleBndTest.java
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-no-resolve/src/main/java/bnd/test/SimpleBndTest.java
@@ -1,0 +1,15 @@
+package bnd.test;
+
+import static org.junit.Assert.assertNotNull;
+import static org.osgi.framework.FrameworkUtil.getBundle;
+
+import org.junit.Test;
+
+public class SimpleBndTest {
+
+	@Test
+	public void test() {
+		assertNotNull(getBundle(SimpleBndTest.class).getBundleContext());
+	}
+	
+}

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-no-resolve/test.bndrun
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-no-resolve/test.bndrun
@@ -1,0 +1,8 @@
+-standalone: target/index.xml
+
+-runfw: 		org.apache.felix.framework;version='[5.6.1,5.6.1]'
+-runee: 		JavaSE-1.7
+
+-runbundles: osgi.enroute.junit.wrapper;version='[4.12.0,4.12.1)',\
+ osgi.enroute.hamcrest.wrapper;version='[1.3.0,1.3.1)',\
+ test-no-resolve;version='[0.0.1,0.0.2)'

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-with-resolve/bnd.bnd
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-with-resolve/bnd.bnd
@@ -1,0 +1,2 @@
+Require-Capability: osgi.extender;filter:="(osgi.extender=osgi.component)"
+Test-Cases: ${classes;CONCRETE;PUBLIC;NAMED;*Test}

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-with-resolve/pom.xml
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-with-resolve/pom.xml
@@ -1,0 +1,57 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+
+	<parent>
+		<groupId>biz.aQute.bnd-test</groupId>
+		<artifactId>test</artifactId>
+		<version>0.0.1</version>
+	</parent>
+
+	<groupId>biz.aQute.bnd-test</groupId>
+	<artifactId>test-with-resolve</artifactId>
+	<packaging>jar</packaging>
+
+	<dependencies>
+		<dependency>
+			<groupId>org.osgi</groupId>
+			<artifactId>org.osgi.core</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.osgi</groupId>
+			<artifactId>osgi.enroute.junit.wrapper</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.osgi</groupId>
+			<artifactId>osgi.enroute.hamcrest.wrapper</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.apache.felix</groupId>
+			<artifactId>org.apache.felix.framework</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.apache.felix</groupId>
+			<artifactId>org.apache.felix.scr</artifactId>
+		</dependency>
+	</dependencies>
+
+	<build>
+		<plugins>
+			<plugin>
+				<groupId>biz.aQute.bnd</groupId>
+				<artifactId>bnd-testing-maven-plugin</artifactId>
+				<configuration>
+					<resolve>true</resolve>
+				</configuration>
+			</plugin>
+			<plugin>
+				<groupId>biz.aQute.bnd</groupId>
+				<artifactId>bnd-indexer-maven-plugin</artifactId>
+			</plugin>
+			<plugin>
+				<groupId>biz.aQute.bnd</groupId>
+				<artifactId>bnd-maven-plugin</artifactId>
+			</plugin>
+		</plugins>
+	</build>
+</project>

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-with-resolve/src/main/java/bnd/test/ResolvingBndTest.java
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-with-resolve/src/main/java/bnd/test/ResolvingBndTest.java
@@ -1,0 +1,33 @@
+package bnd.test;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.osgi.framework.FrameworkUtil.getBundle;
+
+import org.junit.Test;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+
+public class ResolvingBndTest {
+
+	@Test
+	public void test() {
+		assertNotNull(getBundle(ResolvingBndTest.class).getBundleContext());
+	}
+	
+	@Test
+	public void test2() {
+		BundleContext ctx = getBundle(ResolvingBndTest.class).getBundleContext();
+		assertNotNull(ctx);
+		
+		boolean found = false;
+		for (Bundle bundle : ctx.getBundles()) {
+			if("org.apache.felix.scr".equals(bundle.getSymbolicName())) {
+				found = true;
+				break;
+			}
+		}
+		assertTrue(found);
+		
+	}
+}

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-with-resolve/test.bndrun
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/test-with-resolve/test.bndrun
@@ -1,0 +1,12 @@
+-standalone: target/index.xml
+
+-runfw: 		org.apache.felix.framework;version='[5.6.1,5.6.1]'
+-runee: 		JavaSE-1.7
+
+-runrequires: osgi.identity;filter:='(osgi.identity=test-with-resolve)'
+
+-runbundles: \
+ org.apache.felix.scr; version='[2.0.2,2.0.3)',\
+ osgi.enroute.hamcrest.wrapper; version='[1.3.0,1.3.1)',\
+ osgi.enroute.junit.wrapper; version='[4.12.0,4.12.1)',\
+ test-with-resolve; version='[0.0.1,0.0.2)'

--- a/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/verify.groovy
+++ b/maven/bnd-testing-maven-plugin/src/test/resources/integration-test/test/verify.groovy
@@ -1,0 +1,25 @@
+import java.io.*;
+
+println "basedir ${basedir}"
+
+assert new File("${basedir}/build.log").exists();
+
+List<String> fileContents = new ArrayList<String>();
+
+BufferedReader reader = new File("${basedir}/build.log").newReader();
+
+String s = null;
+
+while((s = reader.readLine()) != null) {
+    fileContents.add(s);
+}
+
+// Simple test
+int idx = fileContents.indexOf("Tests run  : 1");
+assert idx != -1;
+assert fileContents.get(idx + 1).equals("Passed     : 1");
+
+// Resolving test
+idx = fileContents.indexOf("Tests run  : 2");
+assert idx != -1;
+assert fileContents.get(idx + 1).equals("Passed     : 2");

